README.md
# FCSA Number

An Angular directive that validates numbers and adds commas for the thousands separator or vice versa depending of the configuration.

So when the user enters `1000` into a textbox and tabs out, the value will be formatted to include the thousands separator and display: `1,000` or `1.000`

If an invalid number is entered, the textbox will be invalid with the `fcsaNumber` error.

## Installation

#### With Bower

    bower install angular-fcsa-number

Then reference `angular-fcsa-number/src/fcsaNumber.js` in your project.

#### Manually

Copy `src/fcsaNumber.js` into your project and reference it.

## Quick Start

Add the `fcsa-number` module as a dependency to you Angular app.
 
    angular.module('yourApp', ['fcsa-number']);

Add the `fcsa-number` attribute to textboxes you want to have validated and formatted with thousands separators.

    <input type='text' ng-model='company.employeeCount' fcsa-number />

When an invalid number is entered by the user, the form and control will become invalid and the 'fcsa-number-invalid' class will be added to the text box.

## Options

Without any options passed to it, fcsa-number will validate that the input is a valid number and will also add commas for the thousand separators.


#### max

Validates the number is not above the max number.

    fcsa-number="{ max: 100 }"

 * Valid: 100
 * Invalid: 101

#### min

Validates the number is not below the min number.

    fcsa-number="{ min: -5 }"

 * Valid: -5
 * Invalid: -6

#### maxDecimals

Validates the number does not have more than the specified number of decimals.

    fcsa-number="{ maxDecimals: 2 }"

 * Valid: 1.23
 * Invalid: 1.234

#### maxDigits

Validates the number does not have more than the specified number of digits.

    fcsa-number="{ maxDigits: 2 }"

 * Valid: 76
 * Invalid: 123

#### preventInvalidInput

By default users are allowed to enter invalid characters, and then the textbox is marked invalid. 
If you want to prevent users from entering invalid characters altogether, then use the `preventInvalidInput` option.
If the user presses the 'a' key, the directive will catch it and prevent 'a' from being shown in the textbox.

    fcsa-number="{ preventInvalidInput: true }"

#### prepend

Prepends the specified text before the number.

    fcsa-number="{ prepend: '$' }"

#### append

Appends the specified text after the number.

    fcsa-number="{ append: '%' }"

## Default Options

It's possible to set the options globally too. You do this by calling `fcsaNumberConfigProvider.setDefaultOptions()` 
inside a config function in your app.

Here's a code example that sets the default options:

```javascript
var app = angular.module('yourApp', ['fcsa-number']);
app.config(['fcsaNumberConfigProvider', function(fcsaNumberConfigProvider) {
  fcsaNumberConfigProvider.setDefaultOptions({
    max: 100,
    min: 0
  });
}]);
```

Here's a code example that sets the thousands separator to ".":

```javascript
   fcsaNumberConfigProvider.setDefaultOptions({
        preventInvalidInput: true,
        autoFormat: true,
        keepThousandsOnInput: true,
        decimalSeparator: ',',
        thousandsSeparator: '.',
        renderOnKeyup: true,
        maxDigits: 24,
    });
```


The default options can be overridden locally by passing in an options object: `fcsa-number="{ max: 10 }"`

## Developing

Grunt is used to compile the CoffeeScript files and run the tests. To get started run the following commands on the command line:

    // installs the required node modules
    npm install

    // installs the required bower components
    bower install

Run the following command to automatically compile and run the unit and end to end tests whenever you make a change to a file.

    grunt karma:unit:start express:dev watch

To just run the protractor tests, you can run the following command.

    grunt express:dev e2e


# FCSA Number (Español)

Una directiva Angular que valida números y agrega comas para el separador de miles o viceversa dependiendo de la conconfiguración.

Por lo tanto, cuando el usuario introduce `1000` en un cuadro de texto y se separa, el valor se formateará para incluir el separador de miles y mostrar:` 1,000` o `1.000`

Si se introduce un número no válido, el cuadro de texto no será válido con el error `fcsaNumber`.

## Instalación

#### Con Bower
    bower install angular-fcsa-number

Luego haga referencia a `angular-fcsa-number / src / fcsaNumber.js` en su proyecto.

#### Manualmente

Copie `src / fcsaNumber.js` en su proyecto y haga referencia a él.

## Inicio rápido

Agregue el módulo `fcsa-number` como una dependencia para usted.

    angular.module ('yourApp', ['fcsa-number']);

Agregue el atributo `fcsa-number` a los cuadros de texto que desee validar y formatear con separadores de miles.

    <input type = 'text' ng-modelo = 'company.employeeCount' fcsa-number />

Cuando el usuario ingresa un número no válido, el formulario y el control se convertirán en no válidos y la clase 'fcsa-number-invalid' se agregará al cuadro de texto.

## Opciones

Sin ninguna opción pasada a ella, fcsa-number validará que la entrada es un número válido y también agregará comas para los mil separadores.


#### max

Valida el número no está por encima del número máximo.

    fcsa-number = "{max: 100}"
    
     * Válido: 100
     * No válido: 101

#### min

Valida que el número no está por debajo del número mínimo.

        fcsa-number = "{min: -5}"
    
     * Válido: -5
     * No válido: -6

#### maxDecimals

Valida el número no tiene más que el número especificado de decimales.

        fcsa-number = "{maxDecimals: 2}"
    
     * Válido: 1.23
     * No válido: 1.234

#### maxDigits

Valida el número no tiene más que el número especificado de dígitos.

        fcsa-number = "{maxDigits: 2}"
    
     * Válido: 76
     * No válido: 123

#### preventInvalidInput

De forma predeterminada, se permite a los usuarios introducir caracteres no válidos y, a continuación, el cuadro de texto se marca como no válido.
Si desea evitar que los usuarios ingresen caracteres inválidos, utilice la opción `preventInvalidInput`.
Si el usuario pulsa la tecla 'a', la directiva lo capturará y evitará que se muestre 'a' en el cuadro de texto.

    fcsa-number = "{preventInvalidInput: true}"

#### prepend

Preinforma el texto especificado antes del número.

    fcsa-number = "{prepend: '$'}"

#### adjuntar

Añade el texto especificado después del número.

    fcsa-number = "{append: '%'}"

## Opciones predeterminadas

También es posible configurar las opciones globalmente. Para ello, llame a `fcsaNumberConfigProvider.setDefaultOptions ()`
dentro de una función de configuración en tu aplicación.

He aquí un ejemplo de código que establece las opciones predeterminadas:


```javascript
var app = angular.module('yourApp', ['fcsa-number']);
app.config(['fcsaNumberConfigProvider', function(fcsaNumberConfigProvider) {
  fcsaNumberConfigProvider.setDefaultOptions({
    max: 100,
    min: 0
  });
}]);
```

He aquí un ejemplo de código que establece el separador de miles en ".":

```javascript
   fcsaNumberConfigProvider.setDefaultOptions({
        preventInvalidInput: true,
        autoFormat: true,
        keepThousandsOnInput: true,
        decimalSeparator: ',',
        thousandsSeparator: '.',
        renderOnKeyup: true,
        maxDigits: 24,
    });
```
Las opciones predeterminadas se pueden sobrescribir localmente pasando un objeto de opciones: `fcsa-number =" {max: 10} "`

## Desarrollando

Grunt se utiliza para compilar los archivos CoffeeScript y ejecutar las pruebas. Para empezar, ejecute los siguientes comandos en la línea de comandos:

    // instala los módulos de nodo requeridos
    npm instalar

    // instala los componentes de bower necesarios
    instalación de bower

Ejecute el siguiente comando para compilar y ejecutar automáticamente la unidad y las pruebas de extremo a extremo siempre que realice un cambio en un archivo.

    grunt karma: unidad: start express: dev watch

Para ejecutar las pruebas del transportador, puede ejecutar el siguiente comando.

    grunt express: dev e2e

CHANGELOG.md
# Changelog

### 1.6.0

* Enhancements
  * Input will be formatted according to the angular locale in use. e.g. if using Spanish decimals will use the "," character and groups will use the "." character.

### 1.5.3

* Bug Fixes
  * Fixed a bug where `-1.2` was being formatted as `-1-1.2`

### 1.5.2

* Bug Fixes
  * `preventInvalidInput` is correctly working again

### 1.5.1

* Bug Fixes
  * Fixed issue where overriding a global option on a specific input element would override that 
  option for all input elements

### 1.5.0

* Enhancements
  * Can set default options with `fcsaNumberConfigProvider.setDefaultOptions()`

### 1.4.1

* Bug Fixes
  * `prepend` and `append` values are removed on focus

### 1.4

* Enhancements
  * Added `prepend` option
  * Added `append` option

### 1.3

* Enhancements
  * Not adding commas to the decimal portion of the number
  * Allowing commas to be entered by the user. Thanks @jbulat
  * A hyphen only value is marked invalid

### 1.2.1

* Refactorings
 * using `angular.element` instead of `$` and therefore no longer requiring jQuery to be used

### 1.2

* Enhancements
 * added 'preventInvalidInput' option

### 1.1

* Enhancements
 * renamed `decimals` option to `maxDecimals`
 * renamed `digits` option to `maxDigits`


test/fcsaNumber.spec.js
(function() {
  describe('fcsaNumber', function() {
    var $compile, $scope, compileForm, form, isValid;
    form = void 0;
    $scope = void 0;
    $compile = void 0;
    beforeEach(module('fcsa-number'));
    beforeEach(inject(function($rootScope, _$compile_) {
      $scope = $rootScope;
      $compile = _$compile_;
      return $scope.model = {
        number: 0
      };
    }));
    compileForm = function(options) {
      if (options == null) {
        options = '{}';
      }
      $compile("<form name='form'><input type='text' name='number' ng-model='model.number' fcsa-number='" + options + "' /></form>")($scope);
      $scope.$digest();
      return $scope.form;
    };
    isValid = function(args) {
      args.options || (args.options = '{}');
      $compile("<form name='form'><input type='text' name='number' ng-model='model.number' fcsa-number='" + args.options + "' /></form>")($scope);
      $scope.$digest();
      $scope.form.number.$setViewValue(args.val);
      return $scope.form.number.$valid;
    };
    describe('with renderOnKeyup', function() {
      return describe('on keyup', function() {
        it('adds commas', function() {
          var e, el;
          $scope.model.number = 1000;
          el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number=\"{renderOnKeyup: true}\" />")($scope);
          el = el[0];
          $scope.$digest();
          angular.element(document.body).append(el);
          angular.element(el).triggerHandler('focus');
          expect(el.value).toBe('1000');
          e = new window.KeyboardEvent('keyup');
          delete e.keyCode;
          Object.defineProperty(e, 'keyCode', {
            'value': 49
          });
          el.dispatchEvent(e);
          return expect(el.value).toBe('1,000');
        });
        return it('retains proper model value', function() {
          var e, el;
          $scope.model.number = 1000;
          el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number=\"{renderOnKeyup: true}\" />")($scope);
          el = el[0];
          $scope.$digest();
          angular.element(document.body).append(el);
          angular.element(el).triggerHandler('focus');
          expect(el.value).toBe('1000');
          e = new window.KeyboardEvent('keyup');
          delete e.keyCode;
          Object.defineProperty(e, 'keyCode', {
            'value': 49
          });
          el.dispatchEvent(e);
          return expect($scope.model.number).toBe(1000);
        });
      });
    });
    describe('on focus', function() {
      return it('removes the commas', function() {
        var el;
        $scope.model.number = 1000;
        el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number />")($scope);
        el = el[0];
        $scope.$digest();
        angular.element(document.body).append(el);
        angular.element(el).triggerHandler('focus');
        return expect(el.value).toBe('1000');
      });
    });
    describe('on blur', function() {
      it('adds commas', function() {
        var el;
        $scope.model.number = 1000;
        el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number />")($scope);
        el = el[0];
        $scope.$digest();
        angular.element(document.body).append(el);
        angular.element(el).triggerHandler('focus');
        angular.element(el).triggerHandler('blur');
        return expect(el.value).toBe('1,000');
      });
      describe('with negative decimal number', function() {
        return it('correctly formats it', function() {
          var el;
          $scope.model.number = -1000.2;
          el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number />")($scope);
          el = el[0];
          $scope.$digest();
          angular.element(document.body).append(el);
          angular.element(el).triggerHandler('focus');
          angular.element(el).triggerHandler('blur');
          return expect(el.value).toBe('-1,000.2');
        });
      });
      return describe('when more than 3 decimals', function() {
        return it('does not add commas to the decimals', function() {
          var el;
          $scope.model.number = 1234.5678;
          el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number />")($scope);
          el = el[0];
          $scope.$digest();
          angular.element(document.body).append(el);
          angular.element(el).triggerHandler('focus');
          angular.element(el).triggerHandler('blur');
          return expect(el.value).toBe('1,234.5678');
        });
      });
    });
    describe('no options', function() {
      it('validates positive number', function() {
        var valid;
        valid = isValid({
          val: '1'
        });
        return expect(valid).toBe(true);
      });
      it('validates positive number with commas', function() {
        var valid;
        valid = isValid({
          val: '1,23'
        });
        return expect(valid).toBe(true);
      });
      it('validates negative number', function() {
        var valid;
        valid = isValid({
          val: '-1'
        });
        return expect(valid).toBe(true);
      });
      it('invalidates hyphen only', function() {
        var valid;
        valid = isValid({
          val: '-'
        });
        return expect(valid).toBe(false);
      });
      it('validates number with decimals', function() {
        var valid;
        valid = isValid({
          val: '1.1'
        });
        return expect(valid).toBe(true);
      });
      it('validates number with decimals and commas', function() {
        var valid;
        valid = isValid({
          val: '1,123,142.1'
        });
        return expect(valid).toBe(true);
      });
      it('validates number while ignoring extra commas', function() {
        var valid;
        valid = isValid({
          val: '1,1,23,1,4,2.1'
        });
        return expect(valid).toBe(true);
      });
      it('invalidates number with multiple decimals', function() {
        var valid;
        valid = isValid({
          val: '1.1.2'
        });
        return expect(valid).toBe(false);
      });
      return it('invalidates non number', function() {
        var valid;
        valid = isValid({
          val: '1a'
        });
        return expect(valid).toBe(false);
      });
    });
    return describe('options', function() {
      describe('max', function() {
        it('validates numbers below or equal to max', function() {
          var valid;
          valid = isValid({
            options: '{ max: 100 }',
            val: '100'
          });
          return expect(valid).toBe(true);
        });
        return it('invalidates numbers above max', function() {
          var valid;
          valid = isValid({
            options: '{ max: 100 }',
            val: '100.1'
          });
          return expect(valid).toBe(false);
        });
      });
      describe('min', function() {
        it('validates numbers not below the min', function() {
          var valid;
          valid = isValid({
            options: '{ min: 0 }',
            val: '0'
          });
          return expect(valid).toBe(true);
        });
        return it('invalidates numbers below the min', function() {
          var valid;
          valid = isValid({
            options: '{ min: 0 }',
            val: '-0.1'
          });
          return expect(valid).toBe(false);
        });
      });
      describe('maxDigits', function() {
        it('validates positive numbers not above number of digits', function() {
          var valid;
          valid = isValid({
            options: '{ maxDigits: 2 }',
            val: '99'
          });
          return expect(valid).toBe(true);
        });
        it('invalidates positive numbers above number of digits', function() {
          var valid;
          valid = isValid({
            options: '{ maxDigits: 2 }',
            val: '999'
          });
          return expect(valid).toBe(false);
        });
        return it('validates negative numbers not above number of digits', function() {
          var valid;
          valid = isValid({
            options: '{ maxDigits: 2 }',
            val: '-99'
          });
          return expect(valid).toBe(true);
        });
      });
      describe('maxDecimals', function() {
        it('validates numbers without more decimals', function() {
          var valid;
          valid = isValid({
            options: '{ maxDecimals: 2 }',
            val: '1.23'
          });
          return expect(valid).toBe(true);
        });
        return it('invalidates numbers with more decimals', function() {
          var valid;
          valid = isValid({
            options: '{ maxDecimals: 2 }',
            val: '1.234'
          });
          return expect(valid).toBe(false);
        });
      });
      describe('prepend', function() {
        it('prepends the value', function() {
          $scope.model.number = 1000;
          form = compileForm("{ prepend: \"$\" }");
          expect(form.number.$viewValue).toBe('$1,000');
          return expect($scope.model.number).toBe(1000);
        });
        return it('removes the prepend value on focus', function() {
          var el;
          $scope.model.number = 1000;
          el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number='{ prepend: \"$\" }' />")($scope);
          el = el[0];
          $scope.$digest();
          angular.element(document.body).append(el);
          angular.element(el).triggerHandler('focus');
          return expect(el.value).toBe('1000');
        });
      });
      return describe('append', function() {
        it('appends the value', function() {
          $scope.model.number = 100;
          form = compileForm("{ append: \"%\" }");
          expect(form.number.$viewValue).toBe('100%');
          return expect($scope.model.number).toBe(100);
        });
        return it('removes the append value on focus', function() {
          var el;
          $scope.model.number = 100;
          el = $compile("<input type='text' name='number' ng-model='model.number' fcsa-number='{ append: \"%\" }' />")($scope);
          el = el[0];
          $scope.$digest();
          angular.element(document.body).append(el);
          angular.element(el).triggerHandler('focus');
          return expect(el.value).toBe('100');
        });
      });
    });
  });

}).call(this);


test/fcsaNumberConfig.spec.js
(function() {
  describe('fcsaNumberConfig', function() {
    var $compile, $scope, isValid;
    $scope = void 0;
    $compile = void 0;
    beforeEach(function() {
      var testModule;
      testModule = angular.module('testModule', []);
      testModule.config(function(fcsaNumberConfigProvider) {
        return fcsaNumberConfigProvider.setDefaultOptions({
          max: 1000,
          min: 0
        });
      });
      module('fcsa-number', 'testModule');
      return inject(function($rootScope, _$compile_) {
        $scope = $rootScope;
        $compile = _$compile_;
        return $scope.model = {
          number: 0
        };
      });
    });
    isValid = function(args) {
      args.options || (args.options = '{}');
      $compile("<form name='form'><input type='text' name='number' ng-model='model.number' fcsa-number='" + args.options + "' /></form>")($scope);
      $scope.$digest();
      $scope.form.number.$setViewValue(args.val);
      return $scope.form.number.$valid;
    };
    it('default max', function() {
      var valid;
      valid = isValid({
        val: '1001'
      });
      return expect(valid).toBe(false);
    });
    it('override max', function() {
      var valid;
      valid = isValid({
        options: '{ max: 10000 }',
        val: '1001'
      });
      return expect(valid).toBe(true);
    });
    it('default min', function() {
      var valid;
      valid = isValid({
        val: '-1'
      });
      return expect(valid).toBe(false);
    });
    return it('override min', function() {
      var valid;
      valid = isValid({
        options: '{ min: -10 }',
        val: '-1'
      });
      return expect(valid).toBe(true);
    });
  });

}).call(this);


karma.conf.js
// Karma configuration
// Generated on Mon Jun 23 2014 13:06:50 GMT-0500 (Central Daylight Time)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      'client/bower_components/angular/angular.js',
      'client/bower_components/angular-mocks/angular-mocks.js',
      'src/fcsaNumber.js',
      'test/fcsaNumber.spec.js',
      'test/fcsaNumberConfig.spec.js'
    ],


    // list of files to exclude
    exclude: [

    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {

    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,
    background: true
  });
};


e2e/web/public/fcsaNumber.js
/*! angular-fcsa-number (version 1.6.0) 2018-03-12 */
(function() {
  var fcsaNumberModule,
    __hasProp = {}.hasOwnProperty;

  fcsaNumberModule = angular.module('fcsa-number', []);

  fcsaNumberModule.directive('fcsaNumber', [
    'fcsaNumberConfig', '$locale', function(fcsaNumberConfig, $locale) {
      var addGroupsToInteger, controlKeys, convertToEnglishNumber, decimalSeparator, defaultOptions, getOptions, hasMultipleDecimals, isNotControlKey, isNotDigit, isNumber, makeIsValid, makeMaxDecimals, makeMaxDigits, makeMaxNumber, makeMinNumber, thousandSeparator;
      defaultOptions = fcsaNumberConfig.defaultOptions;
      decimalSeparator = function() {
        return $locale.NUMBER_FORMATS.DECIMAL_SEP;
      };
      thousandSeparator = function() {
        return $locale.NUMBER_FORMATS.GROUP_SEP;
      };
      getOptions = function(scope) {
        var option, options, value, _ref;
        options = angular.copy(defaultOptions);
        if (scope.options != null) {
          _ref = scope.$eval(scope.options);
          for (option in _ref) {
            if (!__hasProp.call(_ref, option)) continue;
            value = _ref[option];
            options[option] = value;
          }
        }
        return options;
      };
      convertToEnglishNumber = function(val) {
        var decRegEx, thouRegEx;
        decRegEx = new RegExp("\\" + decimalSeparator(), 'g');
        thouRegEx = new RegExp("\\" + thousandSeparator(), 'g');
        val = val.replace(thouRegEx, '');
        val = val.replace(decRegEx, '.');
        return val;
      };
      isNumber = function(val) {
        return !isNaN(parseFloat(val)) && isFinite(val);
      };
      isNotDigit = function(which, maxDecimals, minVal) {
        var decSep, thouSep;
        decSep = decimalSeparator().charCodeAt(0);
        thouSep = thousandSeparator().charCodeAt(0);
        if (which === thouSep) {
          return true;
        }
        if (which === decSep) {
          if (maxDecimals === 0) {
            return true;
          }
          return false;
        }
        if (which === 45 && minVal >= 0) {
          return true;
        }
        return which < 44 || which > 57 || which === 47;
      };
      controlKeys = [0, 8, 13];
      isNotControlKey = function(which) {
        return controlKeys.indexOf(which) === -1;
      };
      hasMultipleDecimals = function(val) {
        return (val != null) && val.toString().split('.').length > 2;
      };
      makeMaxDecimals = function(maxDecimals) {
        var regexString, validRegex;
        if (maxDecimals > 0) {
          regexString = "^-?\\d*\\.?\\d{0," + maxDecimals + "}$";
        } else {
          regexString = "^-?\\d*$";
        }
        validRegex = new RegExp(regexString);
        return function(val) {
          return validRegex.test(val);
        };
      };
      makeMaxNumber = function(maxNumber) {
        return function(val, number) {
          return number <= maxNumber;
        };
      };
      makeMinNumber = function(minNumber) {
        return function(val, number) {
          return number >= minNumber;
        };
      };
      makeMaxDigits = function(maxDigits) {
        var validRegex;
        validRegex = new RegExp("^-?\\d{0," + maxDigits + "}(\\.\\d*)?$");
        return function(val) {
          return validRegex.test(val);
        };
      };
      makeIsValid = function(options) {
        var validations;
        validations = [];
        if (options.maxDecimals != null) {
          validations.push(makeMaxDecimals(options.maxDecimals));
        }
        if (options.max != null) {
          validations.push(makeMaxNumber(options.max));
        }
        if (options.min != null) {
          validations.push(makeMinNumber(options.min));
        }
        if (options.maxDigits != null) {
          validations.push(makeMaxDigits(options.maxDigits));
        }
        return function(val) {
          var i, number, _i, _ref;
          if (!isNumber(val)) {
            return false;
          }
          if (hasMultipleDecimals(val)) {
            return false;
          }
          number = Number(val);
          for (i = _i = 0, _ref = validations.length; 0 <= _ref ? _i < _ref : _i > _ref; i = 0 <= _ref ? ++_i : --_i) {
            if (!validations[i](val, number)) {
              return false;
            }
          }
          return true;
        };
      };
      addGroupsToInteger = function(val) {
        var decLoc, decSep, decimals, groups, thouSep, wholeNumbers;
        decSep = decimalSeparator();
        thouSep = thousandSeparator();
        val = val.replace(/\./g, decSep);
        decLoc = val.indexOf(decSep);
        decimals = decLoc == -1 ? '' : val.substring(decLoc);
        decLoc = decLoc > -1 ? decLoc : val.length;
        wholeNumbers = val.substring(0, decLoc);
        groups = wholeNumbers.replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1' + thouSep);
        return "" + groups + decimals;
      };
      return {
        restrict: 'A',
        require: 'ngModel',
        scope: {
          options: '@fcsaNumber'
        },
        link: function(scope, elem, attrs, ngModelCtrl) {
          var isValid, options;
          options = getOptions(scope);
          isValid = makeIsValid(options);
          ngModelCtrl.$parsers.unshift(function(viewVal) {
            var englishNumber;
            englishNumber = convertToEnglishNumber(viewVal);
            if (isValid(englishNumber) || !englishNumber) {
              ngModelCtrl.$setValidity('fcsaNumber', true);
              return englishNumber;
            } else {
              ngModelCtrl.$setValidity('fcsaNumber', false);
              return void 0;
            }
          });
          ngModelCtrl.$formatters.push(function(val) {
            if ((options.nullDisplay != null) && (!val || val === '')) {
              return options.nullDisplay;
            }
            if ((val == null) || !isValid(val)) {
              return val;
            }
            ngModelCtrl.$setValidity('fcsaNumber', true);
            val = addGroupsToInteger(val.toString());
            if (options.prepend != null) {
              val = "" + options.prepend + val;
            }
            if (options.append != null) {
              val = "" + val + options.append;
            }
            return val;
          });
          elem.on('blur', function() {
            var formatter, viewValue, _i, _len, _ref;
            viewValue = ngModelCtrl.$modelValue;
            if ((viewValue == null) || !isValid(viewValue)) {
              return;
            }
            _ref = ngModelCtrl.$formatters;
            for (_i = 0, _len = _ref.length; _i < _len; _i++) {
              formatter = _ref[_i];
              viewValue = formatter(viewValue);
            }
            ngModelCtrl.$viewValue = viewValue;
            return ngModelCtrl.$render();
          });
          elem.on('focus', function() {
            var thouPattern, val;
            val = elem.val();
            if (options.prepend != null) {
              val = val.replace(options.prepend, '');
            }
            if (options.append != null) {
              val = val.replace(options.append, '');
            }
            thouPattern = new RegExp("\\" + thousandSeparator(), 'g');
            elem.val(val.replace(thouPattern, ''));
            return elem[0].select();
          });
          if (options.preventInvalidInput === true) {
            return elem.on('keypress', function(e) {
              if (isNotDigit(e.which, options.maxDecimals, options.min) && isNotControlKey(e.which)) {
                return e.preventDefault();
              }
            });
          }
        }
      };
    }
  ]);

  fcsaNumberModule.provider('fcsaNumberConfig', function() {
    var _defaultOptions;
    _defaultOptions = {};
    this.setDefaultOptions = function(defaultOptions) {
      return _defaultOptions = defaultOptions;
    };
    this.$get = function() {
      return {
        defaultOptions: _defaultOptions
      };
    };
  });

}).call(this);


e2e/fcsaNumber.e2e.js
(function() {
  describe('fcsaNumber', function() {
    describe('on blur', function() {
      it('adds the thousand separator', function() {
        var input;
        input = element(By.model('amount'));
        input.clear();
        input.sendKeys('1000\t');
        return expect(input.getAttribute('value')).toBe('1,000');
      });
      return it('does not add commas to the decimal portion', function() {
        var input;
        input = element(By.model('amount'));
        input.clear();
        input.sendKeys('1234.5678\t');
        return expect(input.getAttribute('value')).toBe('1,234.5678');
      });
    });
    it('removes the thousand separators on focus', function() {
      var input;
      input = element(By.model('amount'));
      input.clear();
      input.sendKeys('1000\t');
      input.click();
      return expect(input.getAttribute('value')).toBe('1000');
    });
    return describe('with renderOnKeyup', function() {
      return it('adds the thousand separators on keydown', function() {
        var input;
        input = element(By.id('with-render-on'));
        input.clear();
        input.sendKeys('1000');
        return expect(input.getAttribute('value')).toBe('1,000');
      });
    });
  });

}).call(this);
